deploy/06_asset/09_deploy_asset.ts

```diff
@@ -1,21 +1,102 @@
 import {HardhatRuntimeEnvironment} from 'hardhat/types';
-import {DeployFunction} from 'hardhat-deploy/types';
-
-const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
-  const {deployments, getNamedAccounts} = hre;
-  const {deploy} = deployments;
+import fs from 'fs-extra';
+import {DeployFunction, DeploymentSubmission} from 'hardhat-deploy/types';
+import {BigNumber} from '@ethersproject/bignumber';
+import {getAddress} from '@ethersproject/address';
 
+const func: DeployFunction = async function (
+  hre: HardhatRuntimeEnvironment
+): Promise<void> {
+  const {deployments, getNamedAccounts, getChainId, upgrades, ethers} = hre;
   const {deployer} = await getNamedAccounts();
+  const {log} = deployments;
+
+  const chainId = await getChainId();
 
   const sandContract = await deployments.get('Sand');
 
-  await deploy('Asset', {
-    from: deployer,
-    args: [sandContract.address, deployer, deployer],
-    log: true,
-    skipIfAlreadyDeployed: true,
-  });
+  const assetContract = await deployments.getOrNull('Asset');
+  if (assetContract) {
+    log(`Asset already deployed at ${assetContract.address}`);
+    return;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const Asset = await (ethers as any).getContractFactory('Asset', deployer); // TODO check types with hardhat-ethers and hardhat-deploy-ethers, for now use `any`
+
+  // Problem : cannot use deterministic deployment with openzepelin proxies ?
+  const asset = await upgrades.deployProxy(
+    Asset,
+    [sandContract.address, deployer, deployer],
+    {initializer: 'init', unsafeAllowCustomTypes: false}
+  );
+  await asset.deployed();
+  const implementationStorage = await ethers.provider.getStorageAt(
+    asset.address,
+    '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc'
+  );
+  const implementationAddress = getAddress(
+    BigNumber.from(implementationStorage).toHexString()
+  );
+
+  log(
+    `Asset deployed as Proxy at : ${asset.address}, implementation: ${implementationAddress}`
+  );
+
+  const AssetArtifact = await deployments.getExtendedArtifact('Asset');
+  const assetAsDeployment: DeploymentSubmission = {
+    address: asset.address,
+    ...AssetArtifact,
+    // TODO :transactionHash: transactionHash for Proxy deployment
```

Comment:
Is this still needed?

Reply:
Yes, it would be better. this will be future improvement